src/hotspot/share/logging/circularStringBuffer.cpp

```diff
@@ -0,0 +1,219 @@
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "logging/circularStringBuffer.hpp"
+#include "runtime/os.inline.hpp"
+
+// LogDecorator::None applies to 'constant initialization' because of its constexpr constructor.
+const LogDecorations& CircularStringBuffer::None = LogDecorations(
+    LogLevel::Warning, LogTagSetMapping<LogTag::__NO_TAG>::tagset(), LogDecorators::None);
+
+const char* allocation_failure_msg = "Failed to allocate async logging buffer";
+
+#ifdef LINUX
+CircularMapping::CircularMapping(size_t size)
+  : size(size) {
+  assert(is_aligned(size, os::vm_page_size()), "must be");
+  file = tmpfile();
+  if (file == nullptr) {
+    vm_exit_out_of_memory(size, OOM_MMAP_ERROR, "%s", allocation_failure_msg);
+  }
+  const int fd = fileno(file);
+  if (fd == -1) {
+    vm_exit_out_of_memory(size, OOM_MMAP_ERROR, "%s", allocation_failure_msg);
+  }
+  int ret = ftruncate(fd, size);
+  if (ret != 0) {
+    vm_exit_out_of_memory(size, OOM_MMAP_ERROR, "%s", allocation_failure_msg);
+  }
+  buffer = (char*)mmap(nullptr, size * 2, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+  if (buffer == MAP_FAILED) {
+    vm_exit_out_of_memory(size, OOM_MMAP_ERROR, "%s", allocation_failure_msg);
+  }
+  void* mmap_ret = mmap(buffer, size, PROT_READ | PROT_WRITE, MAP_SHARED | MAP_FIXED, fd, 0);
+  if (mmap_ret == MAP_FAILED) {
+    vm_exit_out_of_memory(size, OOM_MMAP_ERROR, "%s", allocation_failure_msg);
+  }
+  mmap_ret = mmap(buffer + size, size, PROT_READ | PROT_WRITE, MAP_SHARED | MAP_FIXED, fd, 0);
+  if (mmap_ret == MAP_FAILED) {
+    vm_exit_out_of_memory(size, OOM_MMAP_ERROR, "%s", allocation_failure_msg);
+  }
+
+  // Success, notify NMT.
+  MemTracker::record_virtual_memory_reserve(buffer, size, CURRENT_PC, mtLogging);
+  MemTracker::record_virtual_memory_commit(buffer, size, CURRENT_PC);
+}
+#else
+CircularMapping::CircularMapping(size_t size)
+  : buffer(nullptr),
+    size(size) {
+  buffer = os::reserve_memory(size, false, mtLogging);
+  if (buffer == nullptr) {
+    vm_exit_out_of_memory(size, OOM_MMAP_ERROR, "%s", allocation_failure_msg);
+  }
+  bool ret = os::commit_memory(buffer, size, false);
+  if (!ret) {
+    vm_exit_out_of_memory(size, OOM_MMAP_ERROR, "%s", allocation_failure_msg);
+  }
+}
+#endif // LINUX
+
+
+CircularStringBuffer::CircularStringBuffer(StatisticsMap& map, PlatformMonitor& stats_lock, size_t size, bool should_stall)
+  : _should_stall(should_stall),
+    _stats(map),
+    _stats_lock(stats_lock),
+    circular_mapping(size),
+    tail(0),
+    head(0) {}
+
+size_t CircularStringBuffer::used() {
+  size_t h = Atomic::load(&head);
+  size_t t = Atomic::load(&tail);
```

Comment:
I'll change them to use `Atomic::store`, but shouldn't releasing the lock cause a full memory fence and therefore prevent the store from being delayed? Or is there something else that I'm missing?

Reply:
To answer your question: Yes, this can be called without either lock being held. The goal is to have a cheaper variant of `await_message`. I renamed this to `maybe_has_message` to indicate that you must check whether a message exists when you take the consumer lock.

Reply:
If read without any lock then the reads could happen around the same time as a store and so before the lock is released and has any memory affects. And technically the release of a lock only synchronizes with the acquire of the same lock by another thread - any memory affects broader than that are just implementation artifacts. So if you want to be strictly correct you need `Atomic::store` to pair with the lock-free `Atomic::load`. (And on most platforms these will just turn into regular loads and stores anyway.)

Reply:
Actually for a lock-free traversal you need `load_acquire` and `release_store` so that other writes to the buffer, and the data therein, are visible. Sorry I should have recalled that sooner.